ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_service.rs

```diff
@@ -294,6 +369,69 @@ impl TableInfoService {
         Ok(())
     }
 
+    async fn snapshot_indexer_async_v2(
+        context: Arc<ApiContext>,
+        indexer_async_v2: Arc<IndexerAsyncV2>,
+        epoch: u64,
+    ) -> anyhow::Result<()> {
+        let chain_id = context.chain_id().id();
+        // temporary path to store the snapshot
+        let snapshot_dir = context
+            .node_config
+            .get_data_dir()
+            .join(snapshot_folder_name(chain_id as u64, epoch));
+        // rocksdb will create a checkpoint to take a snapshot of full db and then save it to snapshot_path
+        indexer_async_v2
+            .create_checkpoint(&snapshot_dir)
+            .context(format!("DB checkpoint failed at epoch {}", epoch))?;
+
+        Ok(())
+    }
+
+    /// Uploads the snapshot to GCS if found.
+    /// 1. If current epoch is backuped, it will skip the backup.
+    /// 2. If the chain id in the backup metadata does not match with the current network, it will panic.
+    /// Not thread safe.
+    /// TODO(larry): improve the error handling.
+    async fn backup_snapshot_if_present(
+        context: Arc<ApiContext>,
+        backup_restore_operator: Arc<GcsBackupRestoreOperator>,
+    ) {
+        let target_snapshot_directory_prefix =
+            snapshot_folder_prefix(context.chain_id().id() as u64);
+        // Scan the data directory to find the latest epoch to upload.
+        let mut epochs_to_backup = vec![];
+        for entry in std::fs::read_dir(context.node_config.get_data_dir()).unwrap() {
+            let entry = entry.unwrap();
+            let path = entry.path();
+            let file_name = path.file_name().unwrap().to_string_lossy();
+            if path.is_dir()
+                && file_name.starts_with(&target_snapshot_directory_prefix)
+                && !file_name.ends_with(".tmp")
```

Comment:
is it safe to assume that snapshot is done if tmp doesn't show up in the filename? cc @grao1991

Reply:
I don't know? @msmouse any idea?

Reply:
What? You mean rocksdb adds a ".tmp" suffix while taking a checkpoint? That's great design (moving / renaming folder should be assumed atomic) :trollface: 

if rocksdb didn't do it, we should do it manually in your use case.

Reply:
Yea then i'll use the `tmp` for now to determine the snapshot is done.